### PR TITLE
T702 Make Asset Name Unique Transaction

### PR DESCRIPTION
**Pull request by Tolsi**

### CODE REVIEW

File: src/main/scala/com/wavesplatform/state2/diffs/AssetTransactionsDiff.scala [hunk 1]
```diff
@@ -69,4 +71,26 @@ object AssetTransactionsDiff {
...
+      case Some(itx) if !(itx.sender equals tx.sender) => Left(TransactionValidationError(tx, "Asset was issued by other address"))
+      case Some(itx) => Right(itx)
+    }
+    issueTxEi.flatMap(itx => {
+      val assetName = EqByteArray(itx.name)
+      if (state.isAssetNameAvailable(assetName)) {
+        val assetId = EqByteArray(tx.assetId)
+        Right(Diff(height = height,
+          tx = tx,
+          portfolios = Map(Account.fromPublicKey(tx.sender.publicKey) -> Portfolio(
```
Comment (ismagin) [inline src/main/scala/com/wavesplatform/state2/diffs/AssetTransactionsDiff.scala]: FYI, the same as `tx.sender.toAccount` also works

File: src/main/scala/com/wavesplatform/state2/diffs/AssetTransactionsDiff.scala [hunk 2]
```diff
@@ -69,4 +71,26 @@ object AssetTransactionsDiff {
...
+            balance = -tx.fee,
+            leaseInfo = LeaseInfo.empty,
+            assets = Map.empty)),
+          assetsWithUniqueNames = Map(assetName -> assetId)
+        ))
+      } else Left(TransactionValidationError(tx, "Asset name is not available"))
```
Comment (ismagin) [inline src/main/scala/com/wavesplatform/state2/diffs/AssetTransactionsDiff.scala]: consider "Asset name has been verified for $..." ?

File: src/main/scala/com/wavesplatform/state2/patch/LeasePatch.scala [hunk 3]
```diff
@@ -6,19 +6,20 @@ import com.wavesplatform.state2.{Diff, LeaseInfo, Portfolio}
 object LeasePatch {
   def apply(s: StateReader): Diff = {
 
-    def invertLeaseInfo(l: LeaseInfo): LeaseInfo = LeaseInfo(-l.leaseIn, -l.leaseOut)
+    def invertLeaseInfo(l: LeaseInfo): LeaseInfo = LeaseInfo(-l.leaseIn, -l.leaseOut )
 
     val portfolioUpd = s.accountPortfolios
-      .collect { case (acc, pf) if pf.leaseInfo != LeaseInfo.empty =>
-        acc -> Portfolio(0, invertLeaseInfo(pf.leaseInfo), Map.empty) }
+      .filter { case (_, pf) => pf.leaseInfo != LeaseInfo.empty }
```
Comment (ismagin) [inline src/main/scala/com/wavesplatform/state2/patch/LeasePatch.scala]: why change `collect` to `filter` and `map`?

File: src/main/scala/scorex/api/http/assets/AssetsApiRoute.scala [hunk 4]
```diff
@@ -141,6 +141,26 @@ case class AssetsApiRoute(settings: RestAPISettings, wallet: Wallet, state: Stat
     processRequest("burn", (b: BurnRequest) => transactionModule.burnAsset(b, wallet))
 
 
+  @Path("/makeUnique")
```
Comment (ismagin) [inline src/main/scala/scorex/api/http/assets/AssetsApiRoute.scala]: let's avoid camelCase and go with `make-unique`

File: src/test/scala/com/wavesplatform/settings/BlockchainSettingsSpecification.scala [hunk 5]
```diff
@@ -70,7 +71,9 @@ class BlockchainSettingsSpecification extends FlatSpec with Matchers {
     settings.functionalitySettings.allowInvalidReissueInSameBlockUntilTimestamp should be(12)
     settings.functionalitySettings.allowCreateAliasTransactionAfterTimestamp should be(13)
     settings.functionalitySettings.allowMultipleLeaseCancelTransactionUntilTimestamp should be(14)
+    settings.functionalitySettings.allowMultipleLeaseCancelTransactionUntilTimestamp should be(14)
```
Comment (ismagin) [inline src/test/scala/com/wavesplatform/settings/BlockchainSettingsSpecification.scala]: duplicated line

File: src/it/scala/com/wavesplatform/it/Node.scala [hunk 6]
```diff
@@ -126,6 +126,9 @@ class Node(config: Config, nodeInfo: NodeInfo, client: AsyncHttpClient, timer: T
   def issue(sourceAddress: String, name: String, description: String, quantity: Long, decimals: Byte, reissuable: Boolean, fee: Long): Future[Transaction] =
     post("/assets/issue", IssueRequest(sourceAddress, name, description, quantity, decimals, reissuable, fee)).as[Transaction]
 
+  def makeUnique(sourceAddress: String, assetId: String, fee: Long): Future[Transaction] =
+    post("/assets/makeUnique", MakeUniqueAssetRequest(sourceAddress, assetId, fee)).as[Transaction]
```
Comment (ismagin) [inline src/it/scala/com/wavesplatform/it/Node.scala]: `make-unique`

File: src/main/scala/com/wavesplatform/state2/diffs/BlockDiffer.scala [hunk 7]
```diff
@@ -33,7 +33,7 @@ object BlockDiffer extends ScorexLogging {
         })
     }
     val feeDiff = Monoid[Diff].combineAll(accountPortfolioFeesMap.map { case (acc, p) =>
-      new Diff(Map.empty, Map(acc -> p), Map.empty, Map.empty, Map.empty, Map.empty, Map.empty)
+      new Diff(Map.empty, Map(acc -> p), Map.empty, Map.empty, Map.empty, Map.empty, Map.empty, Map.empty)
```
Comment (phearnot) [inline src/main/scala/com/wavesplatform/state2/diffs/BlockDiffer.scala]: Remove all empty maps.

**Pull request merged**